File: arduino_ethernet/src/inviscomm/Frame.cpp
#include "inviscomm/Frame.h"

namespace InvisFrame {
uint16_t crc16CcittFalse(const uint8_t* data,size_t len){
    uint16_t crc=0xFFFF;
    for(size_t i=0;i<len;i++){
        crc^=(uint16_t)data[i]<<8;
        for(uint8_t b=0;b<8;b++) crc=(crc&0x8000)?(uint16_t)((crc<<1)^0x1021):(uint16_t)(crc<<1);
    }
    return crc;
}
void encode(uint8_t sender,uint8_t flags,uint32_t position,uint16_t coordinate,uint8_t out[SIZE]){
    out[0]=MAGIC; out[1]=VERSION; out[2]=sender; out[3]=flags;
    out[4]=(uint8_t)(position>>24); out[5]=(uint8_t)(position>>16); out[6]=(uint8_t)(position>>8); out[7]=(uint8_t)position;
    out[8]=(uint8_t)(coordinate>>8); out[9]=(uint8_t)coordinate;
    const uint16_t crc=crc16CcittFalse(out,10); out[10]=(uint8_t)(crc>>8); out[11]=(uint8_t)crc;
}
bool decode(const uint8_t in[SIZE],uint8_t& sender,uint8_t& flags,uint32_t& position,uint16_t& coordinate){
    if(in[0]!=MAGIC||in[1]!=VERSION) return false;
    const uint16_t expected=((uint16_t)in[10]<<8)|in[11]; if(crc16CcittFalse(in,10)!=expected) return false;
    sender=in[2]; flags=in[3];
    position=((uint32_t)in[4]<<24)|((uint32_t)in[5]<<16)|((uint32_t)in[6]<<8)|in[7];
    coordinate=((uint16_t)in[8]<<8)|in[9]; return coordinate<=1023;
}
}

File: micropython/common/frame.py
try:
    import struct
except ImportError:
    import ustruct as struct


MAGIC = 0x49
VERSION = 0x01

NODE_A = 0x01
NODE_B = 0x02

FLAG_NONE = 0x00

FRAME_FORMAT_WITHOUT_CRC = ">BBBBIH"
FRAME_FORMAT = ">BBBBIHH"

FRAME_SIZE = struct.calcsize(FRAME_FORMAT)


class FrameError(Exception):
    pass


def crc16_ccitt(data, initial=0xFFFF):
    """
    CRC-16/CCITT-FALSE:
    poly=0x1021, init=0xFFFF, refin=false, refout=false.
    """
    crc = initial

    for byte in data:
        crc ^= byte << 8

        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF

    return crc


def encode_frame(sender, position, coordinate, flags=FLAG_NONE):
    if sender not in (NODE_A, NODE_B):
        raise ValueError("Identificador de nodo inválido")

    if position < 0 or position > 0xFFFFFFFF:
        raise ValueError("Posición fuera de rango")

    if coordinate < 0 or coordinate > 1023:
        raise ValueError("Coordenada fuera del espacio InvisComm")

    body = struct.pack(
        FRAME_FORMAT_WITHOUT_CRC,
        MAGIC,
        VERSION,
        sender,
        flags,
        position,
        coordinate,
    )

    checksum = crc16_ccitt(body)

    return body + struct.pack(">H", checksum)


def decode_frame(data):
    if not isinstance(data, (bytes, bytearray)):
        raise FrameError("La trama debe ser bytes")

    if len(data) != FRAME_SIZE:
        raise FrameError(
            "Longitud inválida: {} bytes; esperados: {}".format(
                len(data),
                FRAME_SIZE,
            )
        )

    (
        magic,
        version,
        sender,
        flags,
        position,
        coordinate,
        received_crc,
    ) = struct.unpack(FRAME_FORMAT, data)

    if magic != MAGIC:
        raise FrameError("Cabecera MAGIC inválida")

    if version != VERSION:
        raise FrameError("Versión de trama no soportada")

    if sender not in (NODE_A, NODE_B):
        raise FrameError("Identificador de nodo inválido")

    if coordinate > 1023:
        raise FrameError("Coordenada fuera del espacio InvisComm")

    calculated_crc = crc16_ccitt(data[:-2])

    if received_crc != calculated_crc:
        raise FrameError(
            "CRC inválido: recibido=0x{:04X}, calculado=0x{:04X}".format(
                received_crc,
                calculated_crc,
            )
        )

    return {
        "sender": sender,
        "flags": flags,
        "position": position,
        "coordinate": coordinate,
    }
